dashboard/src/main/resources/templates/component.ftl

```diff
@@ -110,8 +110,30 @@
     </#if>
 
     <h2>Dependencies</h2>
-    <pre class="dependency-tree">${dependencyTree}</pre>
-    
+
+    <#macro formatDependencyNode tree currentNode parent>
+      <#-- tree: ListMultimap<DependencyPath, DependencyPath>, currentNode: DependencyPath -->
+      <#if currentNode.size() gt 0>
+        <#-- root node does not have any leaf -->
+        <#if parent.size() gt 0>
+          <#assign label = 'parent: ' + parent.getLeaf() />
+        <#else>
+          <#assign label = 'root' />
+        </#if>
+        <p class="DEPENDENCY_TREE_NODE" title="${label}">${currentNode.getLeaf()}</p>
+      </#if>
+      <ul>
+        <#list tree.get(currentNode) as childNode>
+          <li class="DEPENDENCY_TREE_NODE">
+            <@formatDependencyNode tree childNode currentNode />
+          </li>
+        </#list>
+      </ul>
+    </#macro>
+    <@formatDependencyNode dependencyTree
+        objectConstructor('com.google.cloud.tools.opensource.dependencies.DependencyPath')
```

Comment:
If we passed nulls here and wrote logic to check for that, would it avoid the need to use objectConstructor?

Reply:
`tree.get(null)` would cause problem as the root node is represented by empty DependencyPath.